ambry-commons/src/main/java/com/github/ambry/compression/CompressionFactory.java

```diff
@@ -0,0 +1,130 @@
+ * - Get a registered compression by name.
+ * - Get default compressor.
+ * - Set default compressor.
+ * - Get decompressor by compressed buffer generated by compress().
+ */
+public class CompressionFactory {
+
+  // The static instance of compression factory.
+  private static volatile CompressionFactory defaultFactory;
+
+  /**
+   * Get the shared instance of compression factory.
+   * The shared instance contains a pre-registered set of supported compression algorithms.
+   * @return The shared compression factory instance.
+   */
+  public static CompressionFactory getDefault() {
+    if (defaultFactory != null) return defaultFactory;
+
+    synchronized(CompressionFactory.class) {
+      if (defaultFactory == null) {
+        defaultFactory = new CompressionFactory();
+
+        // ****** WARNING ********
+        // You may register/add more compression services, but DO NOT remove any existing compression service.
+        // Decompression relies on this map to locate the decompression algorithm using the algorithm name.
+        defaultFactory.register(new LZ4Compression());
+        defaultFactory.register(new ZstdCompression());
+      }
+    }
+
+    return defaultFactory;
+  }
+
+  // Map from algorithm name to instance of compression service.
+  private final Map<String, Compression> registeredCompressionMap = new HashMap<>();
+
+  // The default compressor to use to compress new data.
+  private Compression defaultCompressor;
+
+  /**
+   * Register a compression service.  Every compression service has a unique name.
+   * Decompression will use the registered compression services to find the decompression algorithm.
+   *
+   * @param newCompressionService The compression service to register.
+   */
+  public void register(Compression newCompressionService) {
+    if (newCompressionService == null) {
+      throw new IllegalArgumentException("newCompressionService");
+    }
+    registeredCompressionMap.put(newCompressionService.getAlgorithmName(), newCompressionService);
+  }
+
+  /**
+   * To get a compressor by algorithm name.
+   * The application config could call this method to set the default compressor algorithm.
+   *      CompressionFactory factory = CompressionFactory.getDefault();
+   *      Compression compressor = factory.getCompressorByName("LZ4");
+   *      factory.setDefaultCompressor(compressor);
+   *
+   * @param algorithmName Name of the algorithm to get.
+   * @return Instance of the compression algorithm; null if not found.
+   */
+  public Compression getCompressorByName(String algorithmName) {
+    return registeredCompressionMap.get(algorithmName);
+  }
+
+  /**
+   * Set the default compressor to use when compressing data.  The new compressor can be null to clear default.
+   * If a new compressor is specified, it must be registered.
+   * @param newCompressor The new compressor to use.
+   */
+  public void setDefaultCompressor(Compression newCompressor) {
```

Comment:
nit: maybe we can also register the compressor here in the same call.

Reply:
It's debatable.  The API is just setting the default compressor.  If the impl also registers the compressor, it's considered side-effect.  Caller can call always call register() if it wants.